projection/validator_delegation/handle_msg.go

```diff
@@ -0,0 +1,329 @@
+
+	return newShares, nil
+}
+
+func (projection *ValidatorDelegation) handleUndelegate(
+	rdbTxHandle *rdb.Handle,
+	height int64,
+	validatorAddress string,
+	delegatorAddress string,
+	amount coin.Int,
+	completionTime utctime.UTCTime,
+) error {
+
+	shares, err := projection.unbondAmountToShares(
+		rdbTxHandle,
+		height,
+		validatorAddress,
+		delegatorAddress,
+		amount,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.unbondAmountToShares(): %v", err)
+	}
+
+	returnAmount, err := projection.unbond(
+		rdbTxHandle,
+		height,
+		validatorAddress,
+		delegatorAddress,
+		shares,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.unbond(): %v", err)
+	}
+
+	ubd, err := projection.setUnbondingDelegationEntry(
+		rdbTxHandle,
+		delegatorAddress,
+		validatorAddress,
+		height,
+		completionTime,
+		returnAmount,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.setUnbondingDelegationEntry(): %v", err)
+	}
+
+	if err := projection.insertUBDQueue(rdbTxHandle, ubd, completionTime); err != nil {
+		return fmt.Errorf("error projection.insertUBDQueue(): %v", err)
+	}
+
+	return nil
+}
+
+func (projection *ValidatorDelegation) handleRedelegate(
+	rdbTxHandle *rdb.Handle,
+	height int64,
+	blockTime utctime.UTCTime,
+	delegatorAddress string,
+	validatorSrcAddress string,
+	validatorDstAddress string,
+	amount coin.Int,
+) error {
+
+	shares, err := projection.unbondAmountToShares(
+		rdbTxHandle,
+		height,
+		validatorSrcAddress,
+		delegatorAddress,
+		amount,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.unbondAmountToShares(): %v", err)
+	}
+
+	srcValReturnAmount, err := projection.unbond(
+		rdbTxHandle,
+		height,
+		validatorSrcAddress,
+		delegatorAddress,
+		shares,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.unbond(): %v", err)
+	}
+
+	dstValSharesCreated, err := projection.handleDelegate(
+		rdbTxHandle,
+		height,
+		validatorDstAddress,
+		delegatorAddress,
+		srcValReturnAmount,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.handleDelegate(): %v", err)
+	}
+
+	completionTime, completeNow, err := projection.calculateRedelegationCompleteTime(
+		rdbTxHandle,
+		height,
+		blockTime,
+		validatorSrcAddress,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.calculateRedelegationCompleteTime(): %v", err)
+	}
+
+	if completeNow { // no need to create the redelegation object
+		return nil
+	}
+
+	red, err := projection.setRedelegationEntry(
+		rdbTxHandle,
+		delegatorAddress,
+		validatorSrcAddress,
+		validatorDstAddress,
+		height,
+		completionTime,
+		srcValReturnAmount,
+		dstValSharesCreated,
+	)
+	if err != nil {
+		return fmt.Errorf("error projection.setRedelegationEntry(): %v", err)
+	}
+
+	if err := projection.insertRedelegationQueue(rdbTxHandle, red, completionTime); err != nil {
+		return fmt.Errorf("error projection.insertRedelegationQueue(): %v", err)
+	}
+
+	return nil
+}
+
+func (projection *ValidatorDelegation) handleValidatorUnjailed(
```

Comment:
same as above

Reply:
`VALIDATOR_JAILED` is an event introduced by ourselves. Not really an Msg.. Yea the file the is a bit misleading. I will revise that! Thanks!